### PR TITLE
Convert the rest of the visitors to use `VisitorResult`

### PR DESCRIPTION
**Pull request by Jarcho**
Continuing from #121256.

### CODE REVIEW

File: compiler/rustc_type_ir/src/visit.rs [hunk 1]
```diff
@@ -47,6 +47,83 @@ use std::ops::ControlFlow;
 
 use crate::{self as ty, BoundVars, Interner, IntoKind, Lrc, TypeFlags};
 
+/// Similar to the `Try` trait, but also implemented for `()`.
```
Comment (oli-obk) [inline compiler/rustc_type_ir/src/visit.rs:50]: Move this to rustc_data_structures, so you can avoid all the new crate dependencies (even if you have to add some, they'll already be there transitively)

Comment (Jarcho) [reply compiler/rustc_type_ir/src/visit.rs:50]: `rustc_type_ir` is already a transitive dependency for all of them via `rustc_ast` (or `rustc_hir`, which depends on `rustc_ast`).

**Review by oli-obk — CHANGES_REQUESTED** [review]
All the impls lgtm

File: compiler/rustc_type_ir/src/visit.rs (hunk 1 above)
Comment (oli-obk) [reply compiler/rustc_type_ir/src/visit.rs:50]: huh. Why does ast depend on type_ir

Comment (oli-obk) [reply compiler/rustc_type_ir/src/visit.rs:50]: Ah https://github.com/rust-lang/rust/blob/e9f95949138125bb1b98ed213a41b9aa825bacf5/compiler/rustc_ast/Cargo.toml#L17

Comment (oli-obk) [reply compiler/rustc_type_ir/src/visit.rs:50]: Let's avoid adding more direct deps and instead move this new trait and macros to data structures. It's not type system specific after all

Comment (Jarcho) [reply compiler/rustc_type_ir/src/visit.rs:50]: What's with the `nightly` feature on `rustc_type_ir` and all the optional dependencies? Should I just ignore those or make the whole visit module depend on a feature?

Comment (oli-obk) [reply compiler/rustc_type_ir/src/visit.rs:50]: Hmm I see your point. That's annoying. I guess it's the same with the movability and the mutability enums. If you feel really motivated rn, you could split those out into a new crate and get rid of the ast->type_ir dep

Comment (Jarcho) [reply compiler/rustc_type_ir/src/visit.rs:50]: Just pulling them into their own crate is easy. Doing it in a way that's actually nice is probably going to be a fair bit of work.

Comment (oli-obk) [reply compiler/rustc_type_ir/src/visit.rs:50]: The split is done. Please move it to rustc_ast_ir and adjust the dependencies. Where useful, reexports could help reduce the amount of direct dependencies on rustc_ast_ir

File: compiler/rustc_ast/src/visit.rs [hunk 2]
```diff
@@ -15,8 +15,8 @@
 
 use crate::ast::*;
 
-use core::ops::ControlFlow;
-
+use rustc_ast_ir::visit::VisitorResult;
```
Comment (oli-obk) [inline compiler/rustc_ast/src/visit.rs:18]:
```suggestion
pub use rustc_ast_ir::visit::VisitorResult;
```

And then get rid of all the direct dependencies on rustc_ast_ir

Comment (Jarcho) [reply compiler/rustc_ast/src/visit.rs]: Done.

**Comment by oli-obk** [comment]
@bors p=1 bitrotty

please rebase, we'll land with priority next time

**Pull request merged**